[Rebranding] Show Brave scheme instead of Chrome Scheme in password manager PWA

jagadeshjai opened the pull request:
- Show Brave scheme instead of Chrome Scheme in password manager PWA installation prompt and App info bubble.
- Add a util methods to replace chrome with brave scheme in a common directory and use it across in required places..

Outcome:

[brave_password_manager_chrome_to_brave_scheme.webm](https://github.com/brave/brave-core/assets/45998266/098fad17-947e-4037-8d16-ec68c920a84a)


<!-- Add brave-browser issue below that this PR will resolve -->
Resolves https://github.com/brave/brave-browser/issues/38836

<!-- CI-related labels that can be applied to this PR:
* CI/run-audit-deps (1) - check for known npm/cargo vulnerabilities (audit_deps)
* CI/run-network-audit (1) - run network-audit
* CI/run-upstream-tests - run Chromium unit and browser tests on Linux and Windows (otherwise only on Linux)
* CI/run-linux-arm64, CI/run-macos-x64, CI/run-windows-arm64, CI/run-windows-x86 - run builds that would otherwise be skipped
* CI/skip - do not run CI builds (except noplatform)
* CI/skip-linux-x64, CI/skip-android, CI/skip-macos-arm64, CI/skip-ios, CI/skip-windows-x64 - skip CI builds for specific platforms
* CI/skip-upstream-tests - do not run Chromium unit, or browser tests (otherwise only on Linux)
* CI/skip-all-linters - do not run presubmit and lint checks
* CI/storybook-url (1) - deploy storybook and provide a unique URL for each build

(1) applied automatically when some files are changed (see: https://github.com/brave/brave-core/blob/master/.github/labeler.yml)
-->

## Submitter Checklist:

- [ ] I confirm that no [security/privacy review is needed](https://github.com/brave/brave-browser/wiki/Security-reviews) and no other type of reviews are needed, or that I have [requested](https://github.com/brave/reviews/issues/new/choose) them
- [x] There is a [ticket](https://github.com/brave/brave-browser/issues) for my issue
- [ ] Used Github [auto-closing keywords](https://docs.github.com/en/github/managing-your-work-on-github/linking-a-pull-request-to-an-issue) in the PR description above
- [ ] Wrote a good [PR/commit description](https://google.github.io/eng-practices/review/developer/cl-descriptions.html)
- [ ] Squashed any review feedback or "fixup" commits before merge, so that history is a record of what happened in the repo, not your PR
- [ ] Added appropriate labels (`QA/Yes` or `QA/No`; `release-notes/include` or `release-notes/exclude`; `OS/...`) to the associated issue
- [ ] Checked the PR locally:
  * `npm run test -- brave_browser_tests`, `npm run test -- brave_unit_tests` [wiki](https://github.com/brave/brave-browser/wiki/Tests)
  * `npm run presubmit` [wiki](https://github.com/brave/brave-browser/wiki/Presubmit-checks), `npm run gn_check`, `npm run tslint`
- [ ] Ran `git rebase master` (if needed)

## Reviewer Checklist:

- [ ] A security review [is not needed](https://github.com/brave/brave-browser/wiki/Security-reviews), or a link to one is included in the PR description
- [ ] New files have MPL-2.0 license header
- [ ] Adequate test coverage exists to prevent regressions
- [ ] Major classes, functions and non-trivial code blocks are well-commented
- [ ] Changes in component dependencies are properly reflected in `gn`
- [ ] Code follows the [style guide](https://chromium.googlesource.com/chromium/src/+/HEAD/styleguide/c++/c++.md)
- [ ] Test plan is specified in PR before merging

## After-merge Checklist:

- [ ] The associated issue milestone is set to the smallest version that the
  changes has landed on
- [ ] All relevant documentation has been updated, for instance:
  - [ ] https://github.com/brave/brave-browser/wiki/Deviations-from-Chromium-(features-we-disable-or-remove)
  - [ ] https://github.com/brave/brave-browser/wiki/Proxy-redirected-URLs
  - [ ] https://github.com/brave/brave-browser/wiki/Fingerprinting-Protections
  - [ ] https://github.com/brave/brave-browser/wiki/Brave%E2%80%99s-Use-of-Referral-Codes
  - [ ] https://github.com/brave/brave-browser/wiki/Web-Compatibility-Exceptions-in-Brave
  - [ ] https://github.com/brave/brave-browser/wiki/QA-Guide
  - [ ] https://github.com/brave/brave-browser/wiki/P3A

## Test Plan:

jagadeshjai commented:
@bridiver Could you please review this PR, since you worked on the Scheme rebranding in the initial stages?

jagadeshjai commented:
@simonhong Could you please review this PR?

simonhong commented on common/brave_common_utils.h:
```diff
@@ -0,0 +1,19 @@
+inline constexpr char16_t kChromeSchema16[] = u"chrome://";
+inline constexpr char16_t kBraveSchema16[] = u"brave://";
+
+std::u16string ReplaceChromeToBraveScheme(const std::u16string& url_string);
```
No one used this api?

jagadeshjai replied:
As it is a util file, I thought of adding it for future use cases.

simonhong commented on common/brave_common_utils.cc:
```diff
@@ -0,0 +1,29 @@
+
+#include "brave/common/brave_common_utils.h"
+
+#include "base/strings/strcat.h"
+#include "base/strings/string_util.h"
+#include "base/strings/utf_string_conversions.h"
```
Please add only used headers in this file.

simonhong commented on common/BUILD.gn:
```diff
@@ -43,7 +43,18 @@ source_set("channel_info") {
+}
+
 group("common") {
+  deps = [ ":common_utils" ]
```
In chromium, module that shared from browser and renderer both is added to `common` directory.
Instead, I think adding util files under `brave/browser/ui` is more appropriate ad that api is only used in ui module.

jagadeshjai replied:
Okay.

Just curious, where would you suggest placing this if we want to use it in the `component` directory?

simonhong replied:
Well.. if it's shared with component layer also, I think we can have new folder under `brave/components/`.

jagadeshjai replied:
Okay, Thanks!

simonhong commented on common/brave_common_utils.h:
```diff
@@ -0,0 +1,19 @@
+inline constexpr char16_t kBraveSchema16[] = u"brave://";
+
+std::u16string ReplaceChromeToBraveScheme(const std::u16string& url_string);
+void ReplaceChromeToBraveScheme(std::u16string* url_string);
```
It would be great if you add the test code  for this new api.

jagadeshjai commented:
@simonhong I hope I have resolved all your comments. Please take a look. 😄

simonhong commented on browser/ui/common/BUILD.gn:
```diff
@@ -0,0 +1,19 @@
+source_set("common") {
```
Having another `common` dir under `browser` is not our way.
Instead, you can simply add your file in `browser/ui` directly.
and that file name could be `brave_sheme_utils.h`?

jagadeshjai replied:
Yeah, Sure!

simonhong replied:
Thanks @jagadeshjai I'll run CI for this PR 👍🏼

simonhong replied:
https://github.com/brave/brave-core/pull/24459

simonhong commented:
@jagadeshjai CI only complained dependency error.
```
[2024-07-02T12:36:56.370Z] + cd src/brave
[2024-07-02T12:36:56.370Z] + npm run gn_check -- Static --target_arch=x64
[2024-07-02T12:36:56.370Z] 
[2024-07-02T12:36:56.370Z] > brave-core@1.69.66 gn_check
[2024-07-02T12:36:56.370Z] > node ./build/commands/scripts/commands.js gn_check Static --target_arch=x64
[2024-07-02T12:36:56.370Z] 
[2024-07-02T12:36:56.622Z] -------------------------------
[2024-07-02T12:36:56.622Z] /home/ubuntu/workspace/pr-brave-browser-jadadeshjai-fix__chrome_to_brave_scheme-linux-x64/src
[2024-07-02T12:36:56.622Z] > gn check /home/ubuntu/workspace/pr-brave-browser-jadadeshjai-fix__chrome_to_brave_scheme-linux-x64/src/out/Static
[2024-07-02T12:37:03.162Z] ERROR at //brave/browser/ui/brave_scheme_utils_unittest.cc:6:11: Include not allowed.
[2024-07-02T12:37:03.162Z] #include "brave/browser/ui/brave_scheme_utils.h"
[2024-07-02T12:37:03.162Z]           ^------------------------------------
[2024-07-02T12:37:03.162Z] It is not in any dependency of
[2024-07-02T12:37:03.162Z]   //brave/browser/ui:unit_tests
[2024-07-02T12:37:03.162Z] The include file is in the target(s):
[2024-07-02T12:37:03.162Z]   //brave/browser/ui:ui
[2024-07-02T12:37:03.162Z] which should somehow be reachable.
[2024-07-02T12:37:03.162Z] null
[2024-07-02T12:37:03.162Z] null
```

`"//chrome/browser/ui" should be added to test target's deps list.

jagadeshjai commented:
> `"//chrome/browser/ui" should be added to test target's deps list.

Done!

Would you mind letting me know why we should add  `//chrome/browser/ui` instead of `//brave/browser/ui` or `:ui` by adding  the  `:unit_tests` in visibility list of `//brave/browser/ui`?

simonhong commented:
> > `"//chrome/browser/ui" should be added to test target's deps list.
> 
> Done! @simonhong
> 
> Would you mind letting me know why we should add `//chrome/browser/ui` instead of `//brave/browser/ui` or `:ui` by adding the `:unit_tests` in visibility list of `//brave/browser/ui`?

Good question.
We added `//brave/browser/ui`  to `//chrome/browser/ui`'s deps list as we heavily use it from `chrome/browser/ui`.
And also we heavily use `//chrome/browser/ui` from `//brave/browser/ui`.
So we added `//brave/browser/ui` to `//chrome/browser/ui`'s  `allow_circular_includes_from` also.
As `//brave/browser/ui` and `//chrome/browser/ui` both are same layer and refers each other's imples and both are mixed.
we want to use one deps for both. so we use `//chrome/browser/ui` deps for both.

jagadeshjai commented:
> Good question. We added `//brave/browser/ui` to `//chrome/browser/ui`'s deps list as we heavily use it from `chrome/browser/ui`. And also we heavily use `//chrome/browser/ui` from `//brave/browser/ui`. So we added `//brave/browser/ui` to `//chrome/browser/ui`'s `allow_circular_includes_from` also. As `//brave/browser/ui` and `//chrome/browser/ui` both are same layer and refers each other's imples and both are mixed. we want to use one deps for both. so we use `//chrome/browser/ui` deps for both.


Aha! Got it!
Thanks for the detailed explanation :smiley:

simonhong reviewed (approved):
++ 👍🏼 
PTAL @brave/chromium-src-reviewers

simonhong commented on browser/ui/BUILD.gn:
```diff
@@ -1351,8 +1353,8 @@ source_set("ui") {
-    sources = []
-    deps = []
+    sources = [ "brave_scheme_utils_unittest.cc" ]
+    deps = [ "//testing/gtest" ]
```
```suggestion
    deps = [
      "//chrome/browser/ui",
      "//testing/gtest",
     ]
```

jagadeshjai replied:
Sorry, I thought I have already pushed it.

Pushed now, PTAL.

goodov commented on chromium_src/chrome/browser/ui/web_applications/app_browser_controller.h:
```diff
@@ -0,0 +1,23 @@
+#ifndef BRAVE_CHROMIUM_SRC_CHROME_BROWSER_UI_WEB_APPLICATIONS_APP_BROWSER_CONTROLLER_H_
+#define BRAVE_CHROMIUM_SRC_CHROME_BROWSER_UI_WEB_APPLICATIONS_APP_BROWSER_CONTROLLER_H_
+
+#define FormatUrlOrigin                                              \
```
```
#define FormatUrlOrigin(...) \
  FormatUrlOrigin_ChromiumImpl(__VA_ARGS__); \
  static std::u16string FormatUrlOrigin(__VA_ARGS__)
```

goodov commented on chromium_src/chrome/browser/ui/views/web_apps/web_app_views_utils.cc:
```diff
@@ -0,0 +1,26 @@
+    bool is_primary_text) {
+  std::unique_ptr<views::Label> origin_label =
+      CreateOriginLabelFromStartUrl_ChromiumImpl(start_url, is_primary_text);
+  std::u16string label_text = origin_label->GetText();
```
`if (origin_label) { ... }` or `CHECK(origin_label);`

goodov commented on browser/ui/brave_scheme_utils.cc:
```diff
@@ -0,0 +1,24 @@
+constexpr char16_t kBraveSchema16[] = u"brave://";
+
+void ReplaceChromeToBraveScheme(std::u16string* url_string) {
+  if (base::StartsWith(*url_string, kChromeSchema16,
```
`DCHECK(url_string);` or just pass `url_string` by reference.

goodov commented on browser/ui/brave_scheme_utils.h:
```diff
@@ -0,0 +1,15 @@
+#include <string>
+
+namespace brave_utils {
+void ReplaceChromeToBraveScheme(std::u16string* url_string);
```
please add a comment.

goodov commented on chromium_src/chrome/browser/ui/views/tabs/tab_hover_card_bubble_view.cc:
```diff
@@ -25,33 +27,28 @@
+  std::u16string domain = domain_label_->GetText();
+  brave_utils::ReplaceChromeToBraveScheme(&domain);
+
+  domain_label_->SetData({domain, /*is_filename*/ false});
```
`SetData` should only be called if the replace was performed. Pls make `ReplaceChromeToBraveScheme` return `bool` to detect if a replacement was performed.

jagadeshjai commented:
@goodov Thanks for the review!

Resolved all your comments, PTAL.

goodov commented on browser/ui/brave_scheme_utils_unittest.cc:
```diff
@@ -0,0 +1,22 @@
+
+TEST(BraveSchemeUtilsTest, ReplaceChromeToBraveScheme) {
+  std::u16string url_string = u"chrome://settings";
+  brave_utils::ReplaceChromeToBraveScheme(url_string);
```
should also test the return value

jagadeshjai replied:
Done @goodov .

goodov commented on browser/ui/brave_scheme_utils_unittest.cc:
```diff
@@ -0,0 +1,22 @@
+// Copyright (c) 2024 The Brave Authors. All rights reserved.
```
this file is not added to `GN`, presubmit rightfully complains. You can run locally `npm run presubmit` to see the error.

jagadeshjai replied:
I think it got missed in recent merge conflict, will check and update.

jagadeshjai replied:
Yes, I overlooked during conflicts, I have added it now.
@goodov

goodov reviewed (changes requested):
gn changes required

goodov commented on browser/ui/brave_scheme_utils.cc:
```diff
@@ -0,0 +1,25 @@
+
+namespace brave_utils {
+constexpr char16_t kChromeSchema16[] = u"chrome://";
+constexpr char16_t kBraveSchema16[] = u"brave://";
```
wait, these should be either in anonymous namespace or just be a part of the function. Right now these variables can be declared as `extern ...` and be used anywhere, and also they can conflict with a symbol with the same name elsewhere.

jagadeshjai replied:
@goodov Initially I thought of having this header in header like `inline constexpr char16_t kChromeSchema16[] = u"chrome://";` and use it elsewhere when needed from here instead of redefining in places,  Whats your thought on to go with this?

goodov replied:
not sure if these schemes in `char16_t` will be used widely tbh. I'd just keep them here in anonymous ns.

jagadeshjai replied:
Okay! For now I'll keep it in the anonymous namespace.

jagadeshjai commented:
@goodov  Could you please run the CI for this now, I just resolved a conflict from master.

goodov reviewed (approved):
lgtm. this can be merged cc @bsclifton

bsclifton commented:
Nice work, @jagadeshjai! 😄
